spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/FilteringBatchAcknowledgingMessageListenerAdapter.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener.adapter;
+
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import org.springframework.kafka.listener.BatchAcknowledgingMessageListener;
+import org.springframework.kafka.support.Acknowledgment;
+
+/**
+ * A {@link BatchAcknowledgingMessageListener} adapter that implements filter logic via a
+ * {@link RecordFilterStrategy}. Note that any discarded records will be ack'd when the
+ * delegate listener invokes {@code acknowledge()} on the {@link Acknowledgment}.
+ *
+ * @param <K> the key type.
+ * @param <V> the value type.
+ *
+ * @author Gary Russell
+ *
+ */
+public class FilteringBatchAcknowledgingMessageListenerAdapter<K, V> extends AbstractFilteringMessageListener<K, V>
+		implements BatchAcknowledgingMessageListener<K, V> {
+
+	private final BatchAcknowledgingMessageListener<K, V> delegate;
+
+	/**
+	 * Create an instance with the supplied strategy and delegate listener.
+	 * @param delegate the delegate.
+	 * @param recordFilterStrategy the filter.
+	 */
+	public FilteringBatchAcknowledgingMessageListenerAdapter(BatchAcknowledgingMessageListener<K, V> delegate,
+			RecordFilterStrategy<K, V> recordFilterStrategy) {
+		super(recordFilterStrategy);
+		this.delegate = delegate;
+	}
+
+	@Override
+	public void onMessage(List<ConsumerRecord<K, V>> consumerRecords, Acknowledgment acknowledgment) {
+		Iterator<ConsumerRecord<K, V>> iterator = consumerRecords.iterator();
+		while (iterator.hasNext()) {
+			if (filter(iterator.next())) {
+				iterator.remove();
+			}
+		}
+		if (consumerRecords.size() > 0) {
+			this.delegate.onMessage(consumerRecords, acknowledgment);
```

Comment:
Well, if we filter all the records and use `MANUAL` (it is a fact for this listener type), we never acknowledge them.
Should we send an empty list anyway?
Or just `acknowledge()` directly in the `else`?